Introduce Resolver::PathParser

jhawthorn opened the pull request:
The template resolver is supposed to determine the handler, format, and variant from the path in extract_handler_and_format_and_variant.

Previously this behaviour was close but didn't exactly match the behaviour of finding templates, and in some cases (particularly with handlers or formats missing) would return incorrect results. I've added previously failing test for where the variant was misdetected.

This commit introduces Resolver::PathParser, a class which should be able to accurately from any path inside a view directory be able to tell us exactly the prefix, partial, variant, locale, format, variant, and handler of that template.

This works by building a building a regexp from the known handlers and file types. This requires that any resolvers have their cache cleared when new handlers or types are registered (this was already somewhat the requirement, since resolver lookups are cached, but this makes it necessary in more situations).

---

Keen readers will note that there's now two different regexes being made in this file 😓. One for finding files and this new one. In a follow up PR #39362 I intend to replace the find logic to instead use this regex, which will require some extra filtering. This will be possible because we have deprecated `.`s from being part of the action name, making the template names unambiguously refer to these properties.

This could also allow us in the future to offer `did_you_mean` style suggestions on missing templates (I really love these changes we've been seeing), or even parsing the entire `app/views` directory at boot (or on the first `render` call) and warning the user about any template's extensions we don't understand.

Thanks for reading ❤️

jhawthorn committed:
Introduce Resolver::PathParser

The template resolver is supposed to determine the handler, format, and
variant from the path in extract_handler_and_format_and_variant.

Previously this behaviour was close but didn't exactly match the
behaviour of finding templates, and in some cases (particularly with
handlers or formats missing) would return incorrect results.

This commit introduces Resolver::PathParser, a class which should be
able to accurately from any path inside a view directory be able to tell
us exactly the prefix, partial, variant, locale, format, variant, and
handler of that template.

This works by building a building a regexp from the known handlers and
file types. This requires that any resolvers have their cache cleared
when new handlers or types are registered (this was already somewhat the
requirement, since resolver lookups are cached, but this makes it
necessary in more situations).

jhawthorn committed:
Only use valid handlers in resolver_shared_tests

We don't have jbuilder installed (this would work if it was).

jhawthorn committed:
Add tests of resolver variant handling

The template resolver is supposed to determine the handler, format, and
variant from the path in extract_handler_and_format_and_variant.

This commit adds tests for how variants are parsed from the filenames
found in the resolver. It includes two skipped tests which currently fail.

jhawthorn committed:
Mark PathParser as :nodoc:

p8 commented:
😄 I was trying to add DidYouMean for TemplateMissing errors but had a hard time finding suggestions.

kaspth reviewed (approved):
❤️💪

jhawthorn committed:
Clear cache after setting Template::Types delegate

details_cache_key already references Template::Types.symbols and view
resolvers cache based on default_formats and other values. This
previously wasn't an issue because no views had been looked up before
this was set. Now that we are building a regex from the values of
Template::Types.symbols we need to clear cache after changing this
setting.

jhawthorn committed:
Add railties test of custom handlers and formats
